numpy/core/tests/test_multiarray.py

```diff
@@ -3548,21 +3566,114 @@ def test_test_zero_rank(self):
 
 
 class TestPickling(object):
+    def test_highest_available_pickle_protocol(self):
+        try:
+            import pickle5
+        except ImportError:
+            pickle5 = None
+
+        if sys.version_info[:2] >= (3, 8) or pickle5 is not None:
+            assert pickle.HIGHEST_PROTOCOL >= 5
+        else:
+            assert pickle.HIGHEST_PROTOCOL < 5
+
+    @pytest.mark.skipif(pickle.HIGHEST_PROTOCOL >= 5,
+                        reason=('this tests the error messages when trying to'
+                                'protocol 5 although it is not available'))
+    def test_correct_protocol5_error_message(self):
+        array = np.arange(10)
+        f = io.BytesIO()
+
+        if sys.version_info[:2] in ((3, 6), (3, 7)):
+            # For the specific case of python3.6 and 3.7, raise a clear import
+            # error about the pickle5 backport when trying to use protocol 5
+            # without pickle 5
+            with pytest.raises(ImportError):
+                array.dump(f, protocol=5)
+
+            with pytest.raises(ImportError):
+                array.dumps(protocol=5)
+
+            with pytest.raises(ImportError):
+                array.__reduce_ex__(5)
+
+        elif sys.version_info[:2] < (3, 6):
+            # when calling __reduce_ex__ explicitly with protocol=5 on python
+            # version <3.6, instead of raising an AttributeError saying "pickle
+            # module has no attribute PickleBuffer", raise a ValueError
+            # saying that protocol 5 is not available for this python version
+            with pytest.raises(ValueError):
+                array.__reduce_ex__(5)
+
+            with pytest.raises(ValueError):
+                array.dumps(protocol=5)
+
+            with pytest.raises(ValueError):
+                array.dump(f, protocol=5)
+
+    @pytest.mark.skipif(pickle.HIGHEST_PROTOCOL < 5,
+                        reason="requires pickle protocol 5")
+    def test_record_array_with_object_dtype(self):
+        my_object = object()
+
+        arr_with_object = np.array(
+                [(my_object, 1, 2.0)],
+                dtype=[('a', object), ('b', int), ('c', float)])
+        arr_without_object = np.array(
+                [('xxx', 1, 2.0)],
+                dtype=[('a', str), ('b', int), ('c', float)])
+
+        arr_with_object_bytes_string = pickle.dumps(arr_with_object,
+                                                    protocol=5)
+        arr_without_object_bytes_string = pickle.dumps(arr_without_object,
+                                                       protocol=5)
+
+        assert b'frombuffer' in arr_without_object_bytes_string
+        assert b'frombuffer' not in arr_with_object_bytes_string
```

Comment:
This is an implementation detail not worth testing in my opinion. Instead please check that you can loads the pickles and check that the dtypes are correct.